lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,582 @@
+            .mappings
+            .external_to_internal_num
+            .keys()
+            .map(|i| PointIdType::NumId(*i));
+
+        let iter_uuid = self
+            .mappings
+            .external_to_internal_uuid
+            .keys()
+            .map(|i| PointIdType::Uuid(*i));
+        // order is important here, we want to iterate over the u64 ids first
+        Box::new(iter_num.chain(iter_uuid))
+    }
+
+    fn iter_internal(&self) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        Box::new(
+            (0..self.mappings.internal_to_external.len() as PointOffsetType)
+                .filter(move |i| !self.deleted[*i as usize]),
+        )
+    }
+
+    fn iter_from(
+        &self,
+        external_id: Option<PointIdType>,
+    ) -> Box<dyn Iterator<Item = (PointIdType, PointOffsetType)> + '_> {
+        let full_num_iter = || {
+            self.mappings
+                .external_to_internal_num
+                .iter()
+                .map(|(k, v)| (PointIdType::NumId(*k), *v))
+        };
+        let offset_num_iter = |offset: u64| {
+            self.mappings
+                .external_to_internal_num
+                .range(offset..)
+                .map(|(k, v)| (PointIdType::NumId(*k), *v))
+        };
+        let full_uuid_iter = || {
+            self.mappings
+                .external_to_internal_uuid
+                .iter()
+                .map(|(k, v)| (PointIdType::Uuid(*k), *v))
+        };
+        let offset_uuid_iter = |offset: Uuid| {
+            self.mappings
+                .external_to_internal_uuid
+                .range(offset..)
+                .map(|(k, v)| (PointIdType::Uuid(*k), *v))
+        };
+
+        match external_id {
+            None => {
+                let iter_num = full_num_iter();
+                let iter_uuid = full_uuid_iter();
+                // order is important here, we want to iterate over the u64 ids first
+                Box::new(iter_num.chain(iter_uuid))
+            }
+            Some(offset) => match offset {
+                PointIdType::NumId(idx) => {
+                    // Because u64 keys are less that uuid key, we can just use the full iterator for uuid
+                    let iter_num = offset_num_iter(idx);
+                    let iter_uuid = full_uuid_iter();
+                    // order is important here, we want to iterate over the u64 ids first
+                    Box::new(iter_num.chain(iter_uuid))
+                }
+                PointIdType::Uuid(uuid) => {
+                    // if offset is a uuid, we can only iterate over uuids
+                    Box::new(offset_uuid_iter(uuid))
+                }
+            },
+        }
+    }
+
+    fn iter_ids(&self) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        self.iter_internal()
+    }
+
+    /// Creates a flusher function, that writes the deleted points bitvec to disk.
+    fn mapping_flusher(&self) -> Flusher {
+        self.deleted_wrapper.flusher()
```

Comment:
Am I correct that we only flush the deletions here because the mappings are constant?

I'd prefer to add a comment to make this clear:

```suggestion
        // Only flush deletions because mappings are immutable
        self.deleted_wrapper.flusher()
```